Addressed issue #100 and EAP7-1394

jmfinelli opened the pull request:
This PR addresses #100 and [EAP7-1394](https://issues.redhat.com/browse/EAP7-1394).

jmfinelli commented:
Hi @yersan, could you please help me with this? I'm not sure why the GitHub Action is failing...apparently, it requires golang 1.19 but, for some reason, the Action runs golang 1.17 :confused:

yersan commented:
Hi @jmfinelli , it should be fixed with https://github.com/wildfly/wildfly-operator/pull/279

jmfinelli commented:
> Hi @jmfinelli , it should be fixed with #279

Thanks @yersan, I will re-run all checks for this PR :-)

yersan commented:
Hello @jmfinelli , FYI, we have updated the libraries and SDK and adjusted some loggers under the tx tests. This has generated some conflicts with your PR.

jmfinelli commented:
> Hello @jmfinelli , FYI, we have updated the libraries and SDK and adjusted some loggers under the tx tests. This has generated some conflicts with your PR.

Thanks for your message @yersan. Strangely enough, I just rebased my branch on top of `main` and I didn't get any conflict :confused:

yersan commented on api/v1alpha1/wildflyserver_types.go:
```diff
@@ -38,8 +38,10 @@ type WildFlyServerSpec struct {
 	// DisableHTTPRoute disables the creation a route to the HTTP port of the application service (false if omitted)
-	DisableHTTPRoute    bool                     `json:"disableHTTPRoute,omitempty"`
-	StandaloneConfigMap *StandaloneConfigMapSpec `json:"standaloneConfigMap,omitempty"`
+	DisableHTTPRoute bool `json:"disableHTTPRoute,omitempty"`
+	// DeactivateTransactionRecovery disables the process of recoverying transactions (false if omitted)
+	DeactivateTransactionRecovery bool `json:"deactivateTransactionRecovery,omitempty"`
```
Wrong formatting, generally IDEs do this automatically otherwise you should execute `go fmt ./...` before pushing your changes up.

jmfinelli replied:
> Wrong formatting, generally IDEs do this automatically otherwise you should execute `go fmt ./...` before pushing your changes up.

Great to know! I will update the PR with `go fmt ./...`. Thanks Yeray :-)

jmfinelli replied:
Done. Thanks again for the sharing that command :partying_face:

yersan reviewed (changes requested):
The documentation is still missing, in addition, could you also execute `make manifests` to update the `config/crd/bases/wildfly.org_wildflyservers.yaml`? You can ignore the  bundle/manifests one, this is relevant only when we are releasing.

jmfinelli commented on api/v1alpha1/wildflyserver_types.go:
```diff
@@ -38,8 +38,10 @@ type WildFlyServerSpec struct {
-	DisableHTTPRoute    bool                     `json:"disableHTTPRoute,omitempty"`
-	StandaloneConfigMap *StandaloneConfigMapSpec `json:"standaloneConfigMap,omitempty"`
+	DisableHTTPRoute bool `json:"disableHTTPRoute,omitempty"`
+	// DeactivateTransactionRecovery disables the process of recoverying transactions (false if omitted)
+	DeactivateTransactionRecovery bool `json:"deactivateTransactionRecovery,omitempty"`
```
I will update the docs. Thanks Yeray

jmfinelli commented on api/v1alpha1/wildflyserver_types.go:
```diff
@@ -134,8 +134,10 @@ type PodStatus struct {
+	// +kubebuilder:validation:Enum=ACTIVE;SCALING_DOWN_RECOVERY_INVESTIGATION;SCALING_DOWN_RECOVERY_PROCESSING;SCALING_DOWN_RECOVERY_HEURISTICS;SCALING_DOWN_CLEAN
 	State string `json:"state"`
+	// Counts the recovery attempts when there are in-doubt transactions
+	RecoveryCounter int32 `json:"recoveryCounter"`
```
Done

jbliznak commented on config/crd/bases/wildfly.org_wildflyservers.yaml:
```diff
@@ -63,6 +63,10 @@ spec:
                 minItems: 1
                 type: array
                 x-kubernetes-list-type: set
+              deactivateTransactionRecovery:
```
why isn't the same change in the bundle/manifests/wildfly.org_wildflyservers.yaml ?

and why do we even have two files with basically the same content?

jmfinelli replied:
IIRC, this one is generated automatically with `make manifest`

jbliznak commented on bundle/manifests/wildfly.org_wildflyservers.yaml:
```diff
@@ -766,18 +766,25 @@ spec:
+                      - SCALING_DOWN_RECOVERY_HEURISTICS
                       - SCALING_DOWN_CLEAN
                       type: string
+                    recoveryCounter:
+                      description: Counts recovery attempts when there are in-doubt transactions.
+                      format: int32
+                      type: integer
+                      default: 0
+                      minimum: 0
                   required:
                   - name
                   - podIP
                   - state
+                  - recoveryCounter
```
the change in the other yaml is keeping alphabetical order here, does it make sense to keep both files in sync?

jmfinelli replied:
IIRC, `config/crd/bases/wildfly.org_wildflyservers.yaml` is generated automatically with `make manifest`

mchoma commented on api/v1alpha1/wildflyserver_types.go:
```diff
@@ -134,8 +134,10 @@ type PodStatus struct {
+	// +kubebuilder:validation:Enum=ACTIVE;SCALING_DOWN_RECOVERY_INVESTIGATION;SCALING_DOWN_RECOVERY_PROCESSING;SCALING_DOWN_RECOVERY_HEURISTICS;SCALING_DOWN_CLEAN
 	State string `json:"state"`
+	// Counts the recovery attempts when there are in-doubt/heuristic transactions
+	RecoveryCounter int32 `json:"recoveryCounter"`
```
```suggestion
	RecoveryCounter int32 `json:"recoveryCounter",omitempty`
```

Dont we need this. To avoid situation in https://issues.redhat.com/browse/JBEAP-23493? @yersan

jmesnil replied:
+1 should be: 

```
RecoveryCounter int32 `json:"recoveryCounter,omitempty"`
```

jmfinelli replied:
Thanks @mchoma :-) Updated

jmfinelli commented on api/v1alpha1/wildflyserver_types.go:
```diff
@@ -109,22 +111,20 @@ type WildFlyServerStatus struct {
 const (
-	// PodStateActive represents PodStatus.State when pod is active to serve requests
-	// it's connected in the Service load balancer
+	// (PodStatus.State) PodStateActive represents an active pod that is connected to the load balancer Service
```
Thanks

jmfinelli commented on controllers/transaction_recovery.go:
```diff
@@ -31,136 +31,165 @@ const (
 
-func (r *WildFlyServerReconciler) checkRecovery(reqLogger logr.Logger, scaleDownPod *corev1.Pod, w *wildflyv1alpha1.WildFlyServer) (bool, string, error) {
+// Defines a new type to map the result of the checkRecovery routine
+type checkResult int
```
I created a new type to clarify what the numeric values stand for

jmfinelli commented on controllers/transaction_recovery.go:
```diff
@@ -31,136 +31,165 @@ const (
-		return false, retString, nil
+		retString := fmt.Sprintf("The transaction recovery process contains errors in the logs. The recovery will be retried."+
+			"Pod: %v, log line with error '%v'", scaleDownPod, foundLogLine)
+		return genericError, retString, nil
```
The block of `if foundLogLine != ""` doesn't use `err`

jmfinelli commented on controllers/transaction_recovery.go:
```diff
@@ -277,119 +322,129 @@ func (r *WildFlyServerReconciler) processTransactionRecoveryScaleDown(reqLogger
-					scaleDownErrors.Store(scaleDownPodName, recoveryErr)
-					return
+
+				// Running the recovery check twice to discover in-doubt transactions
```
Yep, we need to run the recovery scan twice to make sure that all txns are recovered

jmfinelli commented on doc/apis.adoc:
```diff
@@ -53,6 +53,8 @@ it defaults to false (application image is expected to use WildFly S2I Builder/R
 | `configMaps` | List of ConfigMap names to mount as volumes in the containers. Each config map is mounted as a read-only volume under `/etc/configmaps/<config map name>` | string[] | false
 | `disableHTTPRoute`| Disable the creation a route to the HTTP port of the application service (false if omitted) | bool | false
+| `deactivateTransactionRecovery`| DeactivateTransactionRecovery disables the process of recoverying transactions (false if omitted) | bool | false
+| `recoveryCounter`| Counts the recovery attempts when there are in-doubt and/or heuristic transactions
```
no problem, thanks

yersan commented:
@jmfinelli Could you rebase this one on top of the latest main? thanks!

yersan reviewed (changes requested):
@jmfinelli CRD and generated openapi are out of sync with your changes. You have to sync them, for example, using `make build` before pushing up the PR

jmfinelli commented:
> @jmfinelli CRD and generated openapi are out of sync with your changes. You have to sync them, for example, using `make build` before pushing up the PR

`make build` executed and PR updated. Thanks @yersan!

yersan commented on manifests/wildfly/1.0.0/manifests/wildflyservers.wildfly.org.crd.yaml:
```diff
@@ -1064,18 +1064,25 @@ spec:
                     state:
-                      description: Represent the state of the Pod, it is used especially
-                        during scale down.
+                      description: Represents the state of the Pod, it is used during scale down.
```
This file should not be modified, this is the CDR for 1.0.0, a version already released

jmfinelli replied:
Done. Thanks

yersan commented on api/v1alpha1/wildflyserver_types.go:
```diff
@@ -38,8 +38,10 @@ type WildFlyServerSpec struct {
-	DisableHTTPRoute    bool                     `json:"disableHTTPRoute,omitempty"`
-	StandaloneConfigMap *StandaloneConfigMapSpec `json:"standaloneConfigMap,omitempty"`
+	DisableHTTPRoute bool `json:"disableHTTPRoute,omitempty"`
+	// DeactivateTransactionRecovery disables the process of recoverying transactions (false if omitted)
```
s/recoverying/recovering

jmfinelli replied:
Done. Thanks

yersan commented:
Thanks @jmfinelli
